### ISSUE TITLE
Invalid sourcemaps on v3.0.0 

### ISSUE THREAD

**Issue by robgraeber** [dlmanning/gulp-sass#577]
Hey just a heads up, when I tried upgrading to 3.0.0 I was getting source map errors when trying to load the sourcemaps again downstream.

Example:
```
return gulp.src([
        rootDir + '/styles/sass/*.scss'
    ])
        .pipe(sourcemaps.init())
        .pipe(sass().on('error', sass.logError))
        .pipe(sourcemaps.write('./'))
        .pipe(gulpif(minifyMe, cssMin()))
        .pipe(gulp.dest(sassifiedDir));
```

Then trying to call sourcemaps.init({loadMaps: true}) elsewhere throws:
```
Error: Invalid mapping: {"generated":{"line":8370,"column":7},"source":"app/vendors/bower_components/robo-ui/dist/sass/mixins/_clearfix.scss","original":{"line":3,"column":-38},"name":null}
```

I downgraded to 2.3.2 and it works now, but wanted to let you know.

**Comment by xzyfer** [comment]
Can you please confirm this problem still occurs without the minifying step.

If so, can you please provide a sourcemap from 2.3.2 and 3.0.0 for us to compare and debug?

**Comment by robgraeber** [comment]
Yeah it happens without the minifying

**Comment by patrickbussmann** [comment]
Same issue here. It happens since I updated to v3

**Comment by vincentmorneau** [comment]
Same here too.

**Comment by vallieres** [comment]
Same here. Confirmed with and without compression.

**Comment by jakbob** [comment]
Same here. In my case the row that causes the error is an `&:before` rule inside a mixin.

**Comment by xzyfer** [comment]
Can someone please an example of .scss code that produces the problem so that I can see it?

**Comment by jakbob** [comment]
Ok, I think I have found a reliable way of reproducing the issue. The crash happens if I have a mixin that uses `&:before` AND use that mixin more than five times. The following snippet is the smallest amout of code I could find that would cause the crash.

```
@mixin test-mixin() {
    &:before {
        content: '';
    }
}

.test1 {
    @include test-mixin();
}

.test2 {
    @include test-mixin();
}

.test3 {
    @include test-mixin();
}

.test4 {
    @include test-mixin();
}

.test5 {
    @include test-mixin();
}

.test6 {
    @include test-mixin();
}
```

**Comment by ceph3us** [comment]
It actually appears that any mixin that uses parent-target selectors combined with any same-level selector is affected. I could reproduce this with mixins that contain `&:after`, `&.class` and `&[disabled]` as well. Tested on both NodeJS 7.4.0 on macOS and NodeJS 5.4.1 on Linux with gulp-sass 3.1.0.

**Comment by ceph3us** [comment]
I've created the smallest test case for this I can here:
https://github.com/ceph3us/break-gulp-sass

This example uses concat but it should be a no-op; it's just a sourcemap aware gulp plugin that will hit the error.

**Comment by ceph3us** [comment]
I ran a quick bisect on this test - it appears to be an upstream issue that appeared from node-sass 4.0

**Comment by pete-hotchkiss** [comment]
is the only known fix for this right now to downgrade to 2.3.2?

**Comment by mattbrunetti** [comment]
@pete-hotchkiss That's the solution I'm using for now. I'm subscribed to this issue and hoping that someone will eventually save the day.

**Comment by robgraeber** [comment]
@mattbrunetti be the change you want to see

**Comment by ceph3us** [comment]
As I suspected, this appears to be an upstream issue:
```zsh
cepheus > ~ > gulp-sass > git bisect bad
251fbc836c245aab860586777960c0ce3040ea5e is the first bad commit
commit 251fbc836c245aab860586777960c0ce3040ea5e
Author: xzyfer <xzyfer@gmail.com>
Date:   Sun Dec 11 00:53:56 2016 +1100

    Bump node-sass@v4.0.0

    Get the latest LibSass.

    https://github.com/sass/node-sass/releases/tag/v4.0.0

:100644 100644 77dfb352bfb5262a03299769dfee544da2fedb8a 1fb705f1eb3bda6945dc2a2015d73667220abbb7 M	package.json
```

I'll see if I can bisect this issue in `node-sass` and file an upstream issue with them if I track it down.

**Comment by ceph3us** [comment]
Confirmed this as an upstream issue with LibSass.

**Comment by mattbrunetti** [comment]
Thank you @ceph3us! You're the man!

**Comment by jakbob** [comment]
Amazing. Thank you @ceph3us!

**Comment by YodasWs** [comment]
I'm using gulp-sass v3.1.0 and I'm running into this error:
```
Error: Invalid mapping: {"generated":{"line":1,"column":28604},"source":"min.css","original":{"line":51,"column":-1},"name":null}
```

I've found the problem is this:
```scss
body > nav,
body > header,
main > section {
  &:nth-child(odd) {
    background: white;
  }
}
```

but then it's fine when you take out the ampersand:
```scss
body > nav,
body > header,
main > section {
  :nth-child(odd) {
    background: white;
  }
}
```
But of course that's a very different selector.

It's fine when you have a single selector:
```scss
main > section {
	&:nth-child(odd) {
		background: white;
	}
}
```
But then you have to duplicate everything, even with a mixin&hellip;

**Comment by deseloper** [comment]
@YodasWs I have the same issue with v3.1.0.

**Comment by b4dnewz** [comment]
for me with v 3.1.0 and this code:

```javascript
var options = {outputStyle: 'compressed'};

var adminStyle = gulp.src('./assets/src/scss/admin.scss')
  .pipe(sourcemaps.init())
  .pipe(sass(options))
  .on('error', notify.onError('Error: <%= error.message %>'))
  .pipe(sourcemaps.write('./'))
  .pipe(gulp.dest('./assets/dist/css'));
```

the sourcemap points to `admin.css` file instead mapping the sass files.. I can't figure out why since is a pretty basic code, pretty much like the [example](https://github.com/dlmanning/gulp-sass#source-maps) in the readme

**Comment by os1r1s110** [comment]
The upstream (libsass) related issue's conversation has been blocked... has anyone found a fix yet?

**Comment by betorobson** [comment]
I got the same error with this code:
```css
.a,
.b,
.c,
.d {

	&[layout=row] {

		&.input-nogrow {

			.form-group .form-input-extras {
				flex: none;
			}

		}

	}

}
```

**Comment by phpanos** [comment]
I have the same issue. Reverting to 2.3.2 fixes the problem. *Raising awareness of the issue*

**Comment by joshstrange** [comment]
There have been a number of comments deleted on the linked LibSass and the issue is locked. Not sure if they plan on doing anything about it. I am going to attempt to downgrade gulp-sass.

**Comment by entr** [comment]
There's already a patch https://github.com/sass/libsass/pull/2540 and should be part of the next libsass release.

**Comment by joffff** [comment]
@entr it looks like it's been merged into libsass master, so should this now be resolved in gulp-sass?  Because I'm still experiencing problems with 3.1.0.

**Comment by Blinnikov** [comment]
Looks like it fixed in v3.2.0

**Comment by xzyfer** [comment]
Great to hear.
